### DISCUSSION TITLE
Sending entities to service (NGSI)

### DISCUSSION THREAD

**Discussion by AnzeMarinko** [c0c0n3/kitt4sme.live#170] [Q&A]
Hello @c0c0n3,

When I try to send entity to service (for now nearly equal to Roughnator), I have a few problems.

I ran the same command always with a different URL to send entity to Roughnator simulation, dockerized image, and platform cluster on a virtual machine (multipass). This command is:

```bash 
curl -v <URL> -H 'Content-Type: application/json' -H 'fiware-service: csic' -d '                             
{
  "id": "urn:ngsi-ld:PartProgram:1",
  "type": "PartProgram",
  "file_name": {
    "type": "Text",
    "value": "sample.json"
  },
  "json_data": {
    "type": "Text",
    "value": "[{sample: 1}]"
  }
}
'
```

The command is from [NGSI persistance](https://github.com/c0c0n3/kitt4sme.live/blob/main/docs/ngsi-persistence.md).

I get the message below when running dockerized app using `<URL>=0.0.0.0:8000/updates`:

```bash
*   Trying 0.0.0.0:8000...
* TCP_NODELAY set
* Connected to 0.0.0.0 (127.0.0.1) port 8000 (#0)
> POST /updates HTTP/1.1
> Host: 0.0.0.0:8000
> User-Agent: curl/7.68.0
> Accept: */*
> Content-Type: application/json
> fiware-service: csic
> Content-Length: 206
> 
* upload completely sent off: 206 out of 206 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 422 Unprocessable Entity
< date: Wed, 05 Oct 2022 12:01:53 GMT
< server: uvicorn
< content-length: 88
< content-type: application/json
< 
* Connection #0 to host 0.0.0.0 left intact
{"detail":[{"loc":["body","data"],"msg":"field required","type":"value_error.missing"}]}
```

What is the missing field?

Than, I get the message below when running simulation app (from `kitt4sme.roughnator/tests/sim`) using `<URL>=0.0.0.0:1026/orion/v2/entities?options=upsert`:

```bash
*   Trying 0.0.0.0:1026...
* TCP_NODELAY set
* Connected to 0.0.0.0 (127.0.0.1) port 1026 (#0)
> POST /orion/v2/entities?options=upsert HTTP/1.1
> Host: 0.0.0.0:1026
> User-Agent: curl/7.68.0
> Accept: */*
> Content-Type: application/json
> fiware-service: csic
> Content-Length: 206
> 
* upload completely sent off: 206 out of 206 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 400 Bad Request
< Connection: Keep-Alive
< Content-Length: 88
< Content-Type: application/json
< Fiware-Correlator: a4098baa-44a7-11ed-80cb-0242ac140005
< Date: Wed, 05 Oct 2022 12:17:12 GMT
< 
* Connection #0 to host 0.0.0.0 left intact
{"orionError":{"code":"400","reasonPhrase":"Bad Request","details":"service not found"}}
```

How to access Orion properly to send it an entity?

And similarly: I get the message below when running platform cluster (`kitt4sme.live`) using `<URL>=10.59.26.91:1026/orion/v2/entities?options=upsert`:

```bash
*   Trying 10.59.26.91:1026...
* TCP_NODELAY set
* connect to 10.59.26.91 port 1026 failed: Connection refused
* Failed to connect to 10.59.26.91 port 1026: Connection refused
* Closing connection 0
curl: (7) Failed to connect to 10.59.26.91 port 1026: Connection refused
```

On which port or even better URL does Orion live from outside of the platform? 

Could you answer all three questions, please?
Thank you.

**Comment by c0c0n3** [comment]
Hi @AnzeMarinko :-)

> I get the message below when running dockerized app using <URL>=0.0.0.0:8000/updates:

The command in the persistence document you linked targets the Orion service which has an endpoint to insert or update arbitrary entities:

```console
curl -v kitt4sme.collab-cloud.eu/orion/v2/entities?options=upsert
...
```

Not sure what service you have on `0.0.0.0:8000`, but judging by the output it should be the [Roughnator][roughnator]? Roughnator has an `/updates` endpoint which accepts an Orion entity update notification

- https://github.com/c0c0n3/kitt4sme.roughnator/blob/main/roughnator/main.py#L26

The JSON payload this endpoint expects is quite different than what you POSTed, it looks like this

```json
{ "data": [ {some entity},  {some other entity}, ... ] }
```

where "some entity", "some other entity", etc. are instances of `MachineEntity`

- https://github.com/c0c0n3/kitt4sme.roughnator/blob/main/roughnator/ngsy.py#L102

That was a long way to say if you want to upload an arbitrary entity to the FIWARE stack, the Roughnator `/update` endpoint is the wrong place :-) But I suppose you'd figured that out already since you wrote

> Than, I get the message below when running simulation app (from kitt4sme.roughnator/tests/sim) using
> URL=0.0.0.0:1026/orion/v2/entities?options=upsert:

This should actually have worked with the Roughnator simulator. If you could please give us more details about it so we can debug it

* Did you start the simulator as documented here: https://github.com/c0c0n3/kitt4sme.roughnator#live-simulator?
* What was the exact `curl` command you used, including the JSON entity you POSTed?

> I get the message below when running platform cluster (kitt4sme.live) using
> URL=10.59.26.91:1026/orion/v2/entities?options=upsert

Judging by the error message you get, it looks like your problem could be the external IP address of your cluster isn't 10.59.26.91. You'd have to figure out what's the right IP address to use I suppose. That depends on what your setup is, but if you followed the bootstrap procedure to create a Multipass VM, then you should be able to get that IP address by running `multipass list xxx` where xxx is the name you gave to the kitt4sme VM.



[roughnator]: https://github.com/c0c0n3/kitt4sme.roughnator
